daml-lf/kv-support/src/main/scala/com/daml/lf/kv/transactions/TransactionConversions.scala

```diff
@@ -0,0 +1,92 @@
+// Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.lf.kv.transactions
+
+import com.daml.lf.kv.ConversionError
+import com.daml.lf.transaction.{
+  NodeId,
+  TransactionCoder,
+  TransactionOuterClass,
+  VersionedTransaction,
+}
+import com.daml.lf.value.ValueCoder
+
+import scala.collection.mutable
+import scala.jdk.CollectionConverters._
+import scala.util.{Failure, Success, Try}
+
+object TransactionConversions {
+
+  def encodeTransaction(
+      tx: VersionedTransaction
+  ): Either[ValueCoder.EncodeError, RawTransaction] =
+    TransactionCoder
+      .encodeTransaction(TransactionCoder.NidEncoder, ValueCoder.CidEncoder, tx)
+      .map(transaction => RawTransaction(transaction.toByteString))
+
+  def decodeTransaction(
+      rawTx: RawTransaction
+  ): Either[ConversionError, VersionedTransaction] =
+    Try(TransactionOuterClass.Transaction.parseFrom(rawTx.byteString)) match {
+      case Success(transaction) =>
+        TransactionCoder
+          .decodeTransaction(
+            TransactionCoder.NidDecoder,
+            ValueCoder.CidDecoder,
+            transaction,
+          )
+          .left
+          .map(ConversionError.DecodeError)
+      case Failure(throwable) => Left(ConversionError.ParseError(throwable.getMessage))
+    }
+
+  def encodeTransactionNodeId(nodeId: NodeId): RawTransaction.NodeId =
+    RawTransaction.NodeId(nodeId.index.toString)
+
+  def decodeTransactionNodeId(transactionNodeId: RawTransaction.NodeId): NodeId =
+    NodeId(transactionNodeId.value.toInt)
+
+  def extractTransactionVersion(rawTransaction: RawTransaction): String =
+    TransactionOuterClass.Transaction.parseFrom(rawTransaction.byteString).getVersion
+
+  def extractNodeId(rawTransactionNode: RawTransaction.Node): RawTransaction.NodeId =
+    RawTransaction.NodeId(
+      TransactionOuterClass.Node.parseFrom(rawTransactionNode.byteString).getNodeId
+    )
+
+  def reconstructTransaction(
+      transactionVersion: String,
+      nodesWithIds: Seq[TransactionNodeIdWithNode],
+  ): Either[ConversionError.ParseError, RawTransaction] = {
+    // Reconstruct roots by considering the transaction nodes in order and
+    // marking all child nodes as non-roots and skipping over them.
+    val nonRoots = mutable.HashSet.empty[RawTransaction.NodeId]
+    Try(TransactionOuterClass.Transaction.newBuilder.setVersion(transactionVersion)).toEither
+      .flatMap { transactionBuilder =>
```

Comment:
this looks pretty hard to read. now I would suggest that you either use `traverse` if that works or alternatively maybe consider sticking to the imperative loop and throw an exception that you turn into an `Either` at the boundary.

Reply:
discussed in the comment above (https://github.com/digital-asset/daml/pull/12216#discussion_r773218063)